src/Log/LogModel.php

```diff
@@ -0,0 +1,216 @@
+	 */
+	public function __construct( $type, $message, $category, $source, $migrationId, $context, $logId ) {
+		$this->setType( $type );
+		$this->category    = $category;
+		$this->source      = $source;
+		$this->migrationId = $migrationId;
+		$this->context     = $context;
+		$this->message     = $message;
+		$this->id          = $logId;
+	}
+
+	/**
+	 * Set log type
+	 * If log type is not supported, it will fallback to NOTICE type
+	 *
+	 * @param string $type
+	 */
+	private function setType( $type ) {
+		$this->type = array_key_exists( $type, LogType::getAllTypes() )
+			? $type
+			: LogType::NOTICE;
+	}
+
+	/**
+	 * Set log message
+	 * If not defined, fallback to default value
+	 *
+	 * @param string|null $message
+	 */
+	private function setMessage( $message ) {
+		$this->message = is_null( $message )
+			? esc_html__( 'Something went wrong', 'give' )
+			: $message;
+	}
+
+	/**
+	 * Set log source
+	 * If not defined, fallback to default value
+	 *
+	 * @param string|null $source
+	 */
+	private function setSource( $source ) {
+		$this->source = is_null( $source )
+			? esc_html__( 'Give Core', 'give' )
+			: $source;
+	}
+
+	 /**
+	 * @return int
+	 */
+	public function getId() {
+		return $this->id;
+	}
+
+	/**
+	 * @return string
+	 */
+	public function getType() {
+		return $this->type;
+	}
+
+	/**
+	 * @return string
+	 */
+	public function getCategory() {
+		return $this->category;
+	}
+
+	/**
+	 * @return string
+	 */
+	public function getSource() {
+		return $this->source;
+	}
+
+	/**
+	 * @return string|null
+	 */
+	public function getMigrationId() {
+		return $this->migrationId;
+	}
+
+	/**
+	 * @return string
+	 */
+	public function getMessage() {
+		return $this->message;
+	}
+
+	/**
+	 * @return array
+	 */
+	public function getContext() {
+		return $this->context;
+	}
+
+	/**
+	 * Get context data
+	 *
+	 * @param  bool  $jsonEncode
+	 *
+	 * @return string|array
+	 */
+	public function getData( $jsonEncode = false ) {
+		$data = [
+			'message' => $this->getMessage(),
+			'context' => $this->getContext(),
+		];
+
+		if ( $jsonEncode ) {
+			return json_encode( $data );
+		}
+
+		return $data;
+	}
+
+	/**
+	 * @param string $key
+	 * @param mixed $value
+	 */
+	private function addContext( $key, $value ) {
+		if ( is_array( $value ) || is_object( $value ) ) {
+			$value = print_r( $value, true );
+		}
+		$this->context[ $key ] = $value;
+	}
+
+	/**
+	 * @param string $type
+	 * @param array $args
+	 */
+	public function __call( $type, $args ) {
+		list ( $message, $source, $context ) = array_pad( $args, 3, null );
+
+		$this->setType( $type );
+		$this->setMessage( $message );
+		$this->setSource( $source );
+
+		// Set additional context
+		if ( is_array( $context ) ) {
+			foreach ( $context as $key => $value ) {
+				$this->addContext( $key, $value );
+			}
+		}
+
+		$this->save();
+	}
+
+	/**
+	 * Save log record
+	 */
+	public function save() {
+		give( LogRepository::class )->insertLog( $this );
+	}
```

Comment:
Currently this would insert a new row every time this is called. It should insert or update based on whether it has an ID, and then store the ID if inserted.

Reply:
@JasonTheAdams maybe we should change the database structure so that we don't have to worry about what the `Log ID` value is and should we update the log or insert the new one. 
For example, we could add another unique column ( log_key ) which will be an md5 hash of the log fields, e.g
`md5( log_type, message, migration_id, category, source, data )`. With this approach, the database will do everything for us and the new log will be inserted only if one of these values changes, otherwise it will just update the date column.

Reply:
To be honest that sounds overly clever. I'm not worried about duplicate low rows, and even want there to be duplicates if the same issue is happening over and over so we get a sense of repetition. This is all I was really thinking:
```suggestion
	public function save() {
		/** @var LogRepository $repository */
		$repository = give( LogRepository::class );
		
		if ( $this->getId() ) {
			$this->id = $repository->updateLog( $this );
		} else {
			$this->id = $repository->insertLog( $this );
		}
	}
```

This logic could also be moved to a `LogRepository::save()` method. I'm happy either way.

Reply:
Resolved: a1e0cdd2